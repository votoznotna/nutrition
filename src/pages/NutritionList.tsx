import React, { useEffect, useContext, FC } from 'react'
import gql from 'graphql-tag'
import { useQuery, useMutation } from '@apollo/react-hooks'
import { INutritionContext } from '../types'
import { NewDessertInput } from "./../../__generated__/globalTypes";
import NewDessert from '../components/NewDessert'
import DessertGrid from '../components/DessertGrid'
import Header from '../components/Header'
import ControlBox from '../components/ControlBox'
import Loading from '../components/Loading'
import { NutritionContext } from '../context/NutritionContext'
import { dessertList } from './__generated__/dessertList'
import { ResetDesserts } from './__generated__/ResetDesserts'
import { CreateDessertVariables, CreateDessert } from './__generated__/CreateDessert'
import { DeleteDesserts, DeleteDessertsVariables } from './__generated__/DeleteDesserts'

const DESSERT_DETAILS = gql`
  fragment DessertDetails on Dessert {
    id
    name
    nutritionInfo {
      calories
      fat
      carbs
      protein
    }
    created
  }
`
const GET_DESSERTS = gql`
  query dessertList {
    desserts {
      ...DessertDetails
    }
  }
  ${DESSERT_DETAILS}
`
const DELETE_DESSERTS = gql`
mutation DeleteDesserts($ids: [String!]!) {
  deleteDesserts(ids: $ids)
}
`
const RESET_DESSERTS = gql`
mutation ResetDesserts {
  resetDesserts {
    ...DessertDetails
  }
}
${DESSERT_DETAILS}
`
const CREATE_DESSERT = gql`
  mutation CreateDessert($input: NewDessertInput!) {
    addDessert(input: $input) {
      ...DessertDetails
    }
  }
  ${DESSERT_DETAILS}
`

const NutritionList: FC = () => {

  const {
    state,
    setModal,
    setSelections,
    setSortField,
    setSortAsc,
    setDesserts
  } = useContext<INutritionContext>(NutritionContext)
  const {
    modal,
    selections
  } = state 

  const { data, loading: dessertListLoading, error: dessertListError, refetch: dessertListRefetch } = useQuery<dessertList>(GET_DESSERTS, { fetchPolicy: 'cache-and-network'})

  const [createDessert, {loading: createDessertLoading, error: createDessertError}] = useMutation<CreateDessert, CreateDessertVariables>(CREATE_DESSERT)
  // , {
  //     update(cache, { data: { addDessert } }) {
  //       const { desserts } = cache.readQuery<any>({ query: GET_DESSERTS })
  
  //       cache.writeQuery({
  //         query: CREATE_DESSERT,
  //         data: { desserts: [addDessert, ...desserts] }
  //       })
  //     }
  //   })

  const [deleteDessert, {loading: deleteDessertsLoading, error: deleteDessertsError}] = useMutation<DeleteDesserts, DeleteDessertsVariables>(DELETE_DESSERTS)
  const [resetDessert, {loading: resetDessertLoading, error: resetDessertError}] = useMutation<ResetDesserts>(RESET_DESSERTS, {
    onCompleted: data => {
      if (data) {
        setDesserts(data.resetDesserts)
      }
    }})

  const desserts = data ? data.desserts : null;

  useEffect(() => {
    desserts && setDesserts(desserts)
  }, [desserts, setDesserts])

  const onReset = async () => {
    await resetDessert({
      variables: {}
    })
    setSelections([])
    setSortField('created')
    setSortAsc(false)
  }

  const onSubmitNewDessert = async (input: NewDessertInput) => {
    setModal(false)
    await createDessert({
      variables: {input}
      })
    dessertListRefetch();
      // .then(
      //   res => {
      //     setNewDessert(res.data.addDessert)
      //     sortDesserts()
      //   },
      //   err => {
      //     console.log(`createDessert err: ${JSON.stringify(err, null, '\t')}`)
      //   }
      // )
  }

  const onDeleteDesserts = () => {
    deleteDessert({
      variables: {ids: selections}
    })
    // .then(
    //   res => {
    //     setDeletedDesserts(res.data.deleteDesserts)
    //     sortDesserts()
    //   },
    //   err => {
    //     console.log(`deleteDessert err: ${JSON.stringify(err, null, '\t')}`)
    //   }
    // )
    setSelections([])
    dessertListRefetch()
  }

  const onAddDessert = () => {
    setModal(true)
  }

  if (modal) {
    return (
      <div className="center">
          <NewDessert onAddDessert={onSubmitNewDessert} onCancel={() => setModal(false)} />
      </div>
    )
  }   

  if (dessertListError || createDessertError || deleteDessertsError || resetDessertError) return <p>ERROR</p>

  const loading = dessertListLoading || resetDessertLoading || createDessertLoading || deleteDessertsLoading

  return (
    <>
        <Header onReset={onReset}></Header> 
        <div className="ph2 pv4">
          <ControlBox onAddDessert={onAddDessert} onDeleteDesserts={onDeleteDesserts} />
          {loading && (<Loading />)}
          {!loading && (<DessertGrid />)}
        </div>
    </>
  )
}

export default NutritionList