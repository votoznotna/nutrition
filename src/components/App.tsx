import React from 'react'
import NutritionList from '../pages/NutritionList'

function App() {
  return (
    <>
      <NutritionList />
    </>
  )
}

export default App
