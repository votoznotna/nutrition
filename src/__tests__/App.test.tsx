import React from "react";
import { MockedProvider } from "@apollo/react-testing";
import { render } from "@testing-library/react";
import NutritionProvider from "../context/NutritionContext";
import { GET_DESSERTS } from "../pages/NutritionList";
import App from "../components/App";

import { act } from "react-dom/test-utils";

const wait = (ms = 0) => act(() => new Promise((done) => setTimeout(done, ms)));

const mocks: any[] = [
  {
    request: {
      query: GET_DESSERTS,
    },
    result: {
      data: {
        desserts: [
          {
            id: "fafa6860-d77d-11ea-9788-ad71c48281af",
            name: "KitKat",
            created: "1596674666982",
            nutritionInfo: {
              calories: 518,
              fat: 26,
              carbs: 65,
              protein: 60,
              __typename: "NutritionInfo",
            },
            __typename: "Dessert",
          },
        ],
      },
    },
  },
];

describe("<App/>", () => {
  it("renders page elements as expected", async () => {
    const { getByText, getByTestId, queryByText, container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <NutritionProvider>
          <App />
        </NutritionProvider>
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();

    expect(queryByText("Nutrition List")).toBeTruthy();

    expect(
      Array.from(container.querySelectorAll("button"))
        .map((btn) => btn.textContent?.trim())
        .toString()
    ).toBe("RESET DATA,ADD NEW,DELETE");

    await wait(); // wait untill loading completed

    let nutritionTable = getByTestId("nutrition-table");
    expect(nutritionTable).toBeInTheDocument();

    let tableField = getByText("Dessert (100g serving)");
    expect(tableField).toBeInTheDocument();

    tableField = getByText("Calories");
    expect(tableField).toBeInTheDocument();

    tableField = getByText("Fat (g)");
    expect(tableField).toBeInTheDocument();

    tableField = getByText("Carbs (g)");
    expect(tableField).toBeInTheDocument();

    tableField = getByText("Protein (g)");
    expect(tableField).toBeInTheDocument();

    let addNewButton = getByText("ADD NEW");
    expect(addNewButton).toBeInTheDocument();

    let deleteButton = getByText("DELETE");
    expect(deleteButton).toBeInTheDocument();
  });
});
