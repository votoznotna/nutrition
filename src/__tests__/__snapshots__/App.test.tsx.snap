// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> renders page elements as expected 1`] = `
<div>
  <section
    class="flex items-center justify-between ph3 f2 f1-l"
  >
    <h6
      class="mv0"
    >
      Nutrition List
    </h6>
    <button
      class="f6 f5-l fw6 pa2 br2 b--dark-green white bg-dark-green"
    >
      <i
        aria-hidden="true"
        class="fa fa-refresh mr3"
      >
         
      </i>
      RESET DATA
    </button>
  </section>
  <div
    class="ph2 pv4"
  >
    <section
      class="flex items-center justify-between mb0 w100 mw100 pa3 bg-enable-light-pink"
    >
      <div>
        <label
          class="dark-pink f6 f5-l fw8"
        >
          0 selected
        </label>
      </div>
      <div>
        <button
          class="f6 f5-l b--white fw6 ph3 pv2 green mr2 bg-white"
        >
          <i
            aria-hidden="true"
            class="fa fa-plus mr3"
          >
             
          </i>
          ADD NEW
        </button>
        <button
          class="f6 f5-l fw6 b--disalbe-light-pink pv2 bg-disable-light-pink"
          disabled=""
        >
          <i
            aria-hidden="true"
            class="fa fa-trash-o mr3"
          >
             
          </i>
          DELETE
        </button>
      </div>
    </section>
    <section
      class="flex items-center dark-gray justify-center pt4 f2 f1-l"
    >
      <div
        class="fa-2x"
      >
        <i
          aria-hidden="true"
          class="fa fa-spinner fa-spin"
        />
      </div>
    </section>
  </div>
</div>
`;

exports[`<App/> renders page elements as expected 2`] = `
<div>
  <section
    class="flex items-center justify-between ph3 f2 f1-l"
  >
    <h6
      class="mv0"
    >
      Nutrition List
    </h6>
    <button
      class="f6 f5-l fw6 pa2 br2 b--dark-green white bg-dark-green"
    >
      <i
        aria-hidden="true"
        class="fa fa-refresh mr3"
      >
         
      </i>
      RESET DATA
    </button>
  </section>
  <div
    class="ph2 pv4"
  >
    <section
      class="flex items-center justify-between mb0 w100 mw100 pa3 bg-enable-light-pink"
    >
      <div>
        <label
          class="dark-pink f6 f5-l fw8"
        >
          0 selected
        </label>
      </div>
      <div>
        <button
          class="f6 f5-l b--white fw6 ph3 pv2 green mr2 bg-white"
        >
          <i
            aria-hidden="true"
            class="fa fa-plus mr3"
          >
             
          </i>
          ADD NEW
        </button>
        <button
          class="f6 f5-l fw6 b--disalbe-light-pink pv2 bg-disable-light-pink"
          disabled=""
        >
          <i
            aria-hidden="true"
            class="fa fa-trash-o mr3"
          >
             
          </i>
          DELETE
        </button>
      </div>
    </section>
    <div
      class="overflow-auto"
    >
      <table
        class="f6 f5-l mw-100 w-100"
        data-testid="nutrition-table"
      >
        <thead>
          <tr>
            <th
              class="pl3 tc fw6 bb b--black-20 pv3 bg-white"
            >
              <input
                class="pv3 bb b--black-20 v-mid"
                id="ids"
                type="checkbox"
              />
            </th>
            <th
              class="cursor-pointer sortable-column tc fw6 bb b--black-20 pv3 bg-white"
              data-testid="name"
              id="name"
            >
              Dessert (100g serving)
              <i
                aria-hidden="true"
                class="fa fa-sort ml1 v-mid"
              />
            </th>
            <th
              class="cursor-pointer sortable-column tc fw6 bb b--black-20 pv3 bg-white"
              id="calories"
            >
              Calories
              <i
                aria-hidden="true"
                class="fa fa-sort ml1 v-mid"
              />
            </th>
            <th
              class="cursor-pointer sortable-column tc fw6 bb b--black-20 pv3 bg-white"
              id="fat"
            >
              Fat (g)
              <i
                aria-hidden="true"
                class="fa fa-sort ml1 v-mid"
              />
            </th>
            <th
              class="cursor-pointer sortable-column tc fw6 bb b--black-20 pv3 bg-white"
              id="carbs"
            >
              Carbs (g)
              <i
                aria-hidden="true"
                class="fa fa-sort ml1 v-mid"
              />
            </th>
            <th
              class="cursor-pointer sortable-column tc fw6 bb b--black-20 pv3 bg-white"
              id="protein"
            >
              Protein (g)
              <i
                aria-hidden="true"
                class="fa fa-sort ml1 v-mid"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="bg-double-light-gray lh-copy"
        >
          <tr>
            <td
              class="pl3 tc fw6 pv3 bb b--black-20"
            >
              <input
                class="v-mid"
                id="fafa6860-d77d-11ea-9788-ad71c48281af"
                type="checkbox"
              />
            </td>
            <td
              class="tc fw6 pv3 bb b--black-20"
            >
              KitKat
            </td>
            <td
              class="tc fw6 pv3 bb b--black-20"
            >
              518
            </td>
            <td
              class="tc fw6 pv3 bb b--black-20"
            >
              26
            </td>
            <td
              class="tc fw6 pv3 bb b--black-20"
            >
              65
            </td>
            <td
              class="tc fw6 pv3 bb b--black-20"
            >
              60
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
